### PR TITLE
[ECP-8329] Handle multiple giftcard only payments

### PR DESCRIPTION
**Pull request by khushboo-singhvi**
## Summary
- New Giftcard Payment Method is created
- Handles scenario where the full amount is being payed by Giftcard, then comes in role this new Giftcard Payment Method
- Summary Blocks are updated with correct summary and final values
- Payment Method listing filtering is corrected on Shipping method on cart page

## Tested scenarios

**Fixed issue**:

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
         }
     ],
     "description": "Official Shopware 6 Plugin to connect to Payment Service Provider Adyen",
-    "version": "3.14.2",
+    "version": "3.15.0",
```
Comment (candemiralp) [inline composer.json:9]: Let's not bump the version on this PR.

File: README.md [hunk 2]
```diff
@@ -67,6 +67,7 @@ The plugin integrates card component (Secured Fields) using Adyen Checkout for a
  - Vipps
  - WeChat Pay
  - Open Banking / Pay by Bank
+ - Giftcard
```
Comment (candemiralp) [inline README.md:70]: There is already a line for Giftcards. Could you please remove the redundant line?

File: src/AdyenPaymentShopware6.php [hunk 3]
```diff
@@ -564,6 +568,20 @@ private function updateTo3100(UpdateContext $updateContext): void
         }
     }
 
+    private function updateTo3110(UpdateContext $updateContext): void
```
Comment (candemiralp) [inline src/AdyenPaymentShopware6.php:571]: Could you please rename this method as `updateTo3150()`?

File: src/Resources/app/storefront/src/cart/cart.plugin.js [hunk 4]
```diff
@@ -48,7 +48,7 @@ export default class CartPlugin extends Plugin {
         this.shoppingCartSummaryBlock = DomAccess.querySelectorAll(document, '.checkout-aside-summary-list');
         this.offCanvasSummaryDetails = null;
         this.shoppingCartSummaryDetails = null;
-
+debugger;
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/cart/cart.plugin.js:51]: Could you please remove `debugger`?

File: src/Resources/app/storefront/src/cart/cart.plugin.js [hunk 5]
```diff
@@ -111,6 +111,7 @@ export default class CartPlugin extends Plugin {
         if (this.paymentMethodInstance) {
             this.paymentMethodInstance.unmount();
         }
+        debugger;
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/cart/cart.plugin.js:114]: Could you please remove debugger?

File: src/Resources/app/storefront/src/cart/cart.plugin.js [hunk 6]
```diff
@@ -162,14 +163,15 @@ export default class CartPlugin extends Plugin {
...
-                    this.saveGiftcardStateData(data, this.selectedGiftcard.id);
+                    this.saveGiftcardStateData(data);
                 }
             }.bind(this)
         );
     }
 
     fetchRedeemedGiftcards() {
         this._client.get(`${adyenGiftcardsConfiguration.fetchRedeemedGiftcardsUrl}`, function (response) {
+            debugger;
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/cart/cart.plugin.js:174]: Could you please remove debugger?

File: src/Resources/app/storefront/src/cart/cart.plugin.js [hunk 7]
```diff
@@ -206,6 +208,9 @@ export default class CartPlugin extends Plugin {
 
             });
             //Update calculations
+            debugger;
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/cart/cart.plugin.js:211]: Could you please remove debugger?

File: src/Resources/app/storefront/src/cart/cart.plugin.js [hunk 8]
```diff
@@ -206,6 +208,9 @@ export default class CartPlugin extends Plugin {
...
+            console.log(this.giftcardDiscount);
+            console.log(response.redeemedGiftcards.totalDiscount);
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/cart/cart.plugin.js:213]: Could you please remove console logs?

File: src/Resources/app/storefront/src/checkout/confirm-order.plugin.js [hunk 9]
```diff
@@ -34,17 +34,18 @@ export default class ConfirmOrderPlugin extends Plugin {
     init() {
         this._client = new HttpClient();
         this.selectedAdyenPaymentMethod = this.getSelectedPaymentMethodKey();
+        debugger;
+        console.log(this.selectedAdyenPaymentMethod);
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/checkout/confirm-order.plugin.js:38]: Could you please remove console logs and debugger?

File: src/Resources/app/storefront/src/checkout/confirm-order.plugin.js [hunk 10]
```diff
@@ -63,7 +64,9 @@ export default class ConfirmOrderPlugin extends Plugin {
                 // replaces confirm button with adyen pay button for paywithgoogle, applepay etc.
                 this.initializeCustomPayButton();
             }
-
+debugger;
+            console.log(this.selectedAdyenPaymentMethod);
+            console.log(this.stateData);
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/checkout/confirm-order.plugin.js:69]: Could you please remove console logs and debugger?

File: src/Resources/app/storefront/src/checkout/confirm-order.plugin.js [hunk 11]
```diff
@@ -518,6 +524,11 @@ export default class ConfirmOrderPlugin extends Plugin {
     }
 
     getSelectedPaymentMethodKey() {
+        debugger;
+        console.log( Object.keys(
+            adyenConfiguration.paymentMethodTypeHandlers));
+        console.log(adyenCheckoutOptions.selectedPaymentMethodHandler);
+     console.log(adyenConfiguration.paymentMethodTypeHandlers['scheme']);
```
Comment (candemiralp) [inline src/Resources/app/storefront/src/checkout/confirm-order.plugin.js:531]: Could you please remove console logs?

File: src/PaymentMethods/GiftCardPaymentMethod.php [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return null;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getLogo(): string
+    {
+        return 'giftcard.png';
```
Comment (peterojo) [inline src/PaymentMethods/GiftCardPaymentMethod.php:55]: Do we have this file in the checkoutshopper remote directory?

Comment (peterojo) [reply src/PaymentMethods/GiftCardPaymentMethod.php:55]: Okay we do, nevermind 😄
https://checkoutshopper-live.adyen.com/checkoutshopper/images/logos/medium/giftcard.png

File: src/Subscriber/PaymentSubscriber.php [hunk 13]
```diff
@@ -452,4 +425,46 @@ public function onKernelRequest(RequestEvent $event)
...
+
+        if ($totalGiftcardBalance > 0) {
+            $totalDiscount =  min($totalGiftcardBalance, $amountInMinorUnits);
+        } else {
+            $totalDiscount = 0;
+        }
+
+        return [
+            'giftcardDiscount' => $totalDiscount,
+            'giftcardBalance' => $totalGiftcardBalance,
```
Comment (peterojo) [inline src/Subscriber/PaymentSubscriber.php:453]: I didn't see any usage for `giftcardBalance` left on the frontend. Why not just return `giftcardDiscount`?

File: src/Controller/StoreApi/OrderApi/OrderApiController.php [hunk 14]
```diff
@@ -176,11 +176,7 @@ public function giftcardStateData(SalesChannelContext $context, Request $request
```
Comment (peterojo) [inline src/Controller/StoreApi/OrderApi/OrderApiController.php:176]: Do we still have `paymentMethod`.`type` in the request body [here](https://github.com/Adyen/adyen-shopware6/pull/454/files#diff-bc795603fdb41f8ae6329ac186aa6df57e44d3603ee3700a98c77971b3660389R163)?

File: src/Resources/app/storefront/src/configuration/adyen.js [hunk 15]
```diff
@@ -146,6 +146,7 @@ export default {
         'afterpay_default': 'handler_adyen_afterpaydefaultpaymentmethodhandler',
         'trustly': 'handler_adyen_trustlypaymentmethodhandler',
         'paysafecard': 'handler_adyen_paysafecardpaymentmethodhandler',
+        'giftcard': 'handler_adyen_giftcardpaymentmethodhandler',
         'givex': 'handler_adyen_givexgiftcardpaymentmethodhandler',
```
Comment (peterojo) [inline src/Resources/app/storefront/src/configuration/adyen.js:150]: do we still need all the other giftcards to be in this list? 🤔

File: src/Subscriber/PaymentSubscriber.php (hunk 13 above)
Comment (khushboo-singhvi) [reply src/Subscriber/PaymentSubscriber.php:453]: Hey, We do use it on onShoppingCartLoaded()

File: src/Resources/app/storefront/src/configuration/adyen.js (hunk 15 above)
Comment (khushboo-singhvi) [reply src/Resources/app/storefront/src/configuration/adyen.js:150]: That we are still not sure, so kept it for now. Once we are done with the actual /payments call, then we can decide over it

File: src/Controller/StoreApi/OrderApi/OrderApiController.php (hunk 14 above)
Comment (khushboo-singhvi) [reply src/Controller/StoreApi/OrderApi/OrderApiController.php:176]: Yes, we do have it.

**Review by candemiralp — APPROVED** [review]

**Review by peterojo — APPROVED** [review]

**Pull request merged**